validation/errors_test.go

```diff
@@ -0,0 +1,81 @@
+package validation
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+)
+
+func TestMakeFuncs(t *testing.T) {
+	testCases := []struct {
+		fn       func() *Error
+		expected ErrorType
+	}{
+		{
+			func() *Error { return Invalid("a", "b", "c") },
+			ErrorTypeInvalid,
+		},
+		{
+			func() *Error { return NotSupported("a", "b", nil) },
+			ErrorTypeNotSupported,
+		},
+		{
+			func() *Error { return Duplicate("a", "b") },
+			ErrorTypeDuplicate,
+		},
+		{
+			func() *Error { return NotFound("a", "b") },
+			ErrorTypeNotFound,
+		},
+		{
+			func() *Error { return Required("a", "b") },
+			ErrorTypeRequired,
+		},
+		{
+			func() *Error { return InternalError("a", fmt.Errorf("b")) },
+			ErrorTypeInternal,
+		},
+	}
+
+	for _, testCase := range testCases {
+		err := testCase.fn()
+		if err.Type != testCase.expected {
+			t.Errorf("expected Type %q, got %q", testCase.expected, err.Type)
+		}
+	}
+}
+
+func TestErrorUsefulMessage(t *testing.T) {
+	s := Invalid("foo", "bar", "deet").Error()
+	for _, part := range []string{"foo", "bar", "deet", ErrorTypeInvalid.String()} {
+		if !strings.Contains(s, part) {
+			t.Errorf("error message did not contain expected part '%v'", part)
```

Comment:
This never actually tells you what the error message is, which makes it a little hard to fix.  We already pull in [testify/assert] if you want to use that ­— it does end up including the actual and expected contents in the output. (In this case, probably using `EqualError` or `Contains`, depending on how you want to go.)

[testify/assert]: https://godoc.org/github.com/stretchr/testify/assert